### PR TITLE
Include Slasher Receiving Methods

### PR DESCRIPTION
**Pull request by rauljordan**
This is part of an initiative to lower the diff in feature/slasher by incremental merging into develop. This PR brings over the method which receives attestations and block headers from the beacon node event feed to pre-process them and prepare to detect slashable offenses. It does not run slashing detection on these yet and instead adds a TODO which will be resolved in the next PR. Attestations and blocks are put into a queue when received and popped when processing each slot.

### CODE REVIEW

File: beacon-chain/slasher/receive.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+// Receive beacon blocks from some source event feed,
+func (s *Service) receiveBlocks(ctx context.Context) {
+	sub := s.serviceCfg.BeaconBlockHeadersFeed.Subscribe(s.beaconBlockHeadersChan)
+	defer close(s.beaconBlockHeadersChan)
+	defer sub.Unsubscribe()
+	for {
+		select {
+		case blockHeader := <-s.beaconBlockHeadersChan:
+			signingRoot, err := blockHeader.Header.HashTreeRoot()
```
Comment (terencechain) [inline beacon-chain/slasher/receive.go:54]: No verification is needed here?

File: beacon-chain/slasher/service.go [hunk 2]
```diff
@@ -14,13 +17,20 @@ var log = logrus.WithField("prefix", "slasher")
...
 
 // Service for running slasher mode in a beacon node.
 type Service struct {
-	params     *Parameters
-	serviceCfg *ServiceConfig
+	params                 *Parameters
+	serviceCfg             *ServiceConfig
+	blksQueue              *blocksQueue
+	attsQueue              *attestationsQueue
+	indexedAttsChan        chan *ethpb.IndexedAttestation
```
Comment (terencechain) [inline beacon-chain/slasher/service.go:34]: Will `indexedAttsChan` be used anywhere else other than `receiveAttestations`? 

If not, then I dont think it should be a service struct field

**Review by terencechain — DISMISSED** [review]

File: beacon-chain/slasher/helpers.go [hunk 3]
```diff
@@ -105,6 +106,15 @@ func validateAttestationIntegrity(att *ethpb.IndexedAttestation) bool {
 	return sourceEpoch < targetEpoch
 }
 
+// Validates the signed beacon block header integrity, ensuring we have no nil values.
+func validateBlockHeaderIntegrity(header *ethpb.SignedBeaconBlockHeader) bool {
+	// If a signed block header is malformed, we drop it.
+	if header == nil || header.Header == nil || len(header.Signature) < 96 || bytes.Equal(header.Signature, make([]byte, 96)) {
```
Comment (prestonvanloon) [inline beacon-chain/slasher/helpers.go:112]: This should always be 96 bytes, right?
```suggestion
	if header == nil || header.Header == nil || len(header.Signature) != 96 || bytes.Equal(header.Signature, make([]byte, 96)) {
```

File: beacon-chain/slasher/queue.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+func (q *blocksQueue) push(blk *slashertypes.SignedBlockHeaderWrapper) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	q.items = append(q.items, blk)
+}
+
+func (q *blocksQueue) dequeue() []*slashertypes.SignedBlockHeaderWrapper {
+	q.lock.RLock()
+	defer q.lock.RUnlock()
```
Comment (prestonvanloon) [inline beacon-chain/slasher/queue.go:67]: Wrong lock

File: beacon-chain/slasher/helpers.go (hunk 3 above)
Comment (terencechain) [reply beacon-chain/slasher/helpers.go:112]: Use config value for this `params.BeaconConfig().BLSSignatureLength` ?

Comment (terencechain) [reply beacon-chain/slasher/helpers.go:112]: Sorry I missed this in my review

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
Please fix lock issue

File: beacon-chain/slasher/queue.go (hunk 4 above)
Comment (prestonvanloon) [reply beacon-chain/slasher/queue.go:67]: Use Lock, not RLock

**Review by prestonvanloon — DISMISSED** [review]

**Review by prestonvanloon — APPROVED** [review]

**Review by terencechain — APPROVED** [review]

**Pull request merged**